Add support for RestGetMapping cancellation

fcofdez opened the pull request.

fcofdez committed:
Add support for RestGetMapping cancellation

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

fcofdez committed:
Remove unnecessary base class

fcofdez committed:
Minor change

fcofdez committed:
Ensure tasks are cancelled

fcofdez commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into rest-get-mappings-cancellation

fcofdez commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into rest-get-mappings-cancellation

DaveCTurner reviewed (commented):
Would it be possible also to check for cancellation on each iteration of the loop in `Metadata#findMappings`? Iterating the mappings is pretty expensive, it'd be great if we could abort that loop early where applicable.

fcofdez committed:
Add support for more granular level task cancellation

fcofdez committed:
Merge branch 'rest-get-mappings-cancellation' of github.com:fcofdez/elasticsearch into rest-get-mappings-cancellation

fcofdez committed:
Merge remote-tracking branch 'origin/master' into rest-get-mappings-cancellation

fcofdez commented:
> Would it be possible also to check for cancellation on each iteration of the loop in Metadata#findMappings? Iterating the mappings is pretty expensive, it'd be great if we could abort that loop early where applicable.

I've added support for a more fine grained task cancellation checks, but I run into some issues while testing it. 
I added a fake `MapperPlugin` that allowed injecting a custom `getFieldFilter` that's used during the get mapping transport action. The fake `getFieldFilter` blocks the thread until some conditions are meet, the problem comes when the http connection shares the same event-loop thread that the blocked transport action. In that scenario the closed connection event is not executed until the action is unblocked, meaning that the task is not cancelled. I've decided to move the `TransportGetMappingsAction` to the `GENERIC` thread pool, but I'm not sure if that change could cause problems elsewhere. Let me know what you think @DaveCTurner. Maybe there's a better way to test it.

DaveCTurner commented:
I think it would be acceptable just to extend `MetadataTests#testFindMappings` to verify that `Metadata#findMappings` checks for cancellation the right number of times. I suppose that wouldn't cover us against passing the wrong thing (e.g. a no-op check) into `findMappings` but I think we can take that risk.

Why is the check a `Consumer<String>` rather than just a `Runnable`? I didn't see anything using the index name, although I didn't check very thoroughly.

fcofdez commented:
> I think it would be acceptable just to extend MetadataTests#testFindMappings to verify that Metadata#findMappings checks for cancellation the right number of times. I suppose that wouldn't cover us against passing the wrong thing (e.g. a no-op check) into findMappings but I think we can take that risk.

Thanks for the feedback, that seems like a good idea 👍 

> Why is the check a Consumer<String> rather than just a Runnable? I didn't see anything using the index name, although I didn't check very thoroughly.

That was just in case it could be useful in the future for other cases, but I'll use a `Runnable` instead.

fcofdez committed:
Simplify tests

fcofdez committed:
Merge remote-tracking branch 'origin/master' into rest-get-mappings-cancellation

fcofdez committed:
Merge remote-tracking branch 'origin/master' into rest-get-mappings-cancellation

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:
```diff
@@ -374,6 +375,21 @@ public boolean equalsAliases(Metadata other) {
      */
     public ImmutableOpenMap<String, MappingMetadata> findMappings(String[] concreteIndices,
```
I think we should inline this overload, there's only a couple of other usages of this method in production code and it's a pretty small thing for the callers to pass in a no-op check if that's what they want.

fcofdez replied:
Addressed in https://github.com/elastic/elasticsearch/pull/72234/commits/e2d96baf2c381300d66cbc415623f23dc93f6bc1

DaveCTurner commented on server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetMappingAction.java:
```diff
@@ -73,14 +75,15 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
-                threadPool.executor(ThreadPool.Names.MANAGEMENT).execute(
+        final HttpChannel httpChannel = request.getHttpChannel();
+        return channel -> new RestCancellableNodeClient(client, httpChannel).admin().indices().getMappings(getMappingsRequest,
+            new RestActionListener<>(channel) {
```
This is quite similar to `DispatchingRestToXContentListener`, with deficiencies on both sides:

- here we have a timeout but `DispatchingRestToXContentListener` doesn't
- `DispatchingRestToXContentListener` checks for cancellation after dispatch but here we don't
- Here we don't have a proper `ToXContentObject`, we call `startObject()` and `endObject()` ourselves, but `DispatchingRestToXContentListener` requires a `ToXContentObject`.

What do you think about moving to `DispatchingRestToXContentListener` here? I think it'd just need a little adapter object to implement the timeout and the outer `XContent` object wrapper, see e.g. `RestClusterStateResponse`.

fcofdez replied:
Addressed in https://github.com/elastic/elasticsearch/pull/72234/commits/e2d96baf2c381300d66cbc415623f23dc93f6bc1

DaveCTurner reviewed (commented):
Looks good, I left a couple more style suggestions.

fcofdez committed:
Move to DispatchingRestToXContentListener

fcofdez commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into rest-get-mappings-cancellation

DaveCTurner reviewed (approved):
LGTM
